### PR TITLE
Fix unreleased compatibility workflow

### PR DESCRIPTION
**Pull request by chatton**
## Description

This PR fixes the unreleased workflow and makes updates to ensure all tests pass.

closes: #XXXX

Workflow run [here](https://github.com/cosmos/ibc-go/actions/runs/5068461482)

Note: the transfer failures should be resolved in [this PR](https://github.com/cosmos/ibc-go/pull/3558/files)

### Commit Message / Changelog Entry

N/A

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: .github/workflows/e2e-compatibility-unreleased.yaml [hunk 1]
```diff
@@ -1,21 +1,48 @@
 name: Compatibility E2E (Unreleased)
-on:
-  workflow_dispatch:
-    inputs:
-      release-branch:
-        description: 'Select unreleased to run all release branches against each other'
-        required: true
-        type: choice
-        options:
-          - unreleased
```
Comment (chatton) [inline .github/workflows/e2e-compatibility-unreleased.yaml:10]: this was a legacy from when the workflows were combined, it's not required anymore

File: .github/workflows/e2e-compatibility-unreleased.yaml [hunk 2]
```diff
@@ -1,21 +1,48 @@
...
+      - name: Build image
+        run: |
+          docker_tag="$(echo ${{ matrix.release-branch }} | sed 's/\//-/')"
+          # TODO: IBC_GO_VERSION does not yet do anything in the tests but is required.
+          docker build . -t "${REGISTRY}/${ORG}/${IMAGE_NAME}:$docker_tag" --build-arg IBC_GO_VERSION=${{ matrix.release-branch }}
+          docker push "${REGISTRY}/${ORG}/${IMAGE_NAME}:$docker_tag"
+      - name: Display image details
+        run: |
+          docker_tag="$(echo ${{ matrix.release-branch }} | sed 's/\//-/')"
+          docker inspect "${REGISTRY}/${ORG}/${IMAGE_NAME}:$docker_tag"
```
Comment (chatton) [inline .github/workflows/e2e-compatibility-unreleased.yaml:44]: explicitly built the release branch images before running tests

File: .github/compatibility-test-matrices/unreleased/ica.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
-  "chain-a": ["release-v7.1.x", "release-v7.0.x", "release-v6.2.x", "release-v6.1.x", "release-v5.3.x", "release-v5.2.x", "release-v4.4.x", "release-v4.3.x", "release-v4.2.x"],
-  "chain-b": ["release-v7.1.x", "release-v7.0.x", "release-v6.2.x", "release-v6.1.x", "release-v5.3.x", "release-v5.2.x", "release-v4.4.x", "release-v4.3.x", "release-v4.2.x"],  
+  "chain-a": ["release-v7.1.x", "release-v7.0.x", "release-v6.2.x", "release-v6.1.x"],
+  "chain-b": ["release-v7.1.x", "release-v7.0.x", "release-v6.2.x", "release-v6.1.x"],
```
Comment (chatton) [inline .github/compatibility-test-matrices/unreleased/ica.json:3]: we don't want to run message server ICA tests on versions that expect intertx

File: .github/compatibility-test-matrices/unreleased/incentivized-transfer-1.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  "chain-b": ["release-v7.1.x", "release-v7.0.x", "release-v6.2.x", "release-v6.1.x", "release-v5.3.x", "release-v5.2.x", "release-v4.4.x", "release-v4.3.x", "release-v4.2.x"],
+  "entrypoint": ["TestIncentivizedTransferTestSuite"],
+  "test": [
+    "TestMsgPayPacketFee_AsyncSingleSender_Succeeds",
+    "TestMsgPayPacketFee_InvalidReceiverAccount"
+  ],
+  "relayer-type": ["rly"],
+  "chain-binary": ["simd"],
+  "chain-image": ["ghcr.io/cosmos/ibc-go-simd"]
+}
```
Comment (chatton) [inline .github/compatibility-test-matrices/unreleased/incentivized-transfer-1.json:12]: we exceeded the 256 job limit, so I split up transfer jobs into 3 parts

File: e2e/testsuite/tx.go [hunk 5]
```diff
@@ -97,9 +97,9 @@ func containsMessage(s string, messages []string) bool {
 
 // AssertTxFailure verifies that an sdk.TxResponse has failed.
 func (s *E2ETestSuite) AssertTxFailure(resp sdk.TxResponse, expectedError *errorsmod.Error) {
+	// In older versions, the codespace and abci codes were different. So in compatibility tests
+	// we can not make assertions on them.
 	errorMsg := fmt.Sprintf("%+v", resp)
-	s.Require().Equal(expectedError.ABCICode(), resp.Code, errorMsg)
-	s.Require().Equal(expectedError.Codespace(), resp.Codespace, errorMsg)
```
Comment (damiannolan) [inline e2e/testsuite/tx.go:102]: could the arg now be built-in `error`? happy to keep it this way

**Review by damiannolan — APPROVED** [review]
Beautiful! Thanks @chatton ❤️

File: e2e/testsuite/tx.go (hunk 5 above)
Comment (colin-axner) [reply e2e/testsuite/tx.go:102]: It'd be nice to be able to keep this functionality if possible. Could we skip only for compatibility? Maybe we could comment them out for now?

**Review by colin-axner — APPROVED** [review]

**Comment by colin-axner** [comment]
Thank you @chatton! :magic_wand:

**Comment by colin-axner** [comment]
> @colin-axner I'm happy to leave them commented out and then open up a follow up issue to figure out a clean way of bypassing them for compatibility tests.

This would be great! :heart:

File: e2e/testsuite/tx.go (hunk 5 above)
Comment (colin-axner) [reply e2e/testsuite/tx.go]: Opened #3652

**Comment by chatton** [comment]
created #3653

**Pull request merged**